docs/source/torch.compiler_troubleshooting.rst

```diff
@@ -1,724 +1,1117 @@
+
+**Contents**
+
+.. contents::
+    :local:
+
+Setting Expectations
+~~~~~~~~~~~~~~~~~~~~
+
+``torch.compile`` is designed as a general-purpose PyTorch compiler.
+Unlike the previous compiler solution, TorchScript, ``torch.compile``
+requires fewer code changes, meaning models typically don't need to be rewritten from scratch.
+It also manages unsupported code more gracefully - unsupported code results in a lost optimization opportunity rather than a crash.
+
+In the ideal world, one can simply apply ``torch.compile`` to any PyTorch model and enjoy automatic speedups.
+However, in reality, code complexities can lead to one of three scenarios:
+
+1. ``torch.compile`` works seamlessly, providing speedups.
+2. Some code modifications are necessary. ``torch.compile`` doesn't crash or take too long,
+   but you might not be seeing significant performance gains.
+3. Extensive changes to your code are required.
+
+We anticipate most code will fall under scenarios (1) and (2).
+This document provides tips, arranged by level of involvement, to help address code issues in scenario (2).
+
+Compile times
+-------------
+
+``torch.compile`` functions as a just-in-time compiler, so the initial one or two runs
+of the compiled function are expected to be significantly slower. Recompilations, which can occur under certain conditions (detailed below),
+will also make runs slower. Various ``torch.compile`` components cache results to
+reduce compilation time for future invocations, even in different processes.
+Cold-start (uncached) compilation time typically ranges from seconds to minutes for common or benchmarked models.
+Larger models may take upwards of 30 minutes to a few hours.
+
+Terminology
+~~~~~~~~~~~
+
+The following terms are relevant to troubleshooting ``torch.compile`` problems.
+
+Graph break
+-----------
+
+``torch.compile`` traces your code and attempts to capture your PyTorch code into a
+single computation graph of PyTorch operators (FX graph). However, this is not always possible.
+When encountering code that can't be traced, a "graph break" occurs.
+A graph break involves compiling the FX graph has been determined so far, running the unsupported code,
+then resuming tracing after the unsupported code with a new FX graph.
+Because the computation graph is broken up, we lose optimization opportunities,
+so model code should avoid graph breaks whenever possible.
+Graph breaks occur on things like:
+
+- Data-dependent if-statements
+- Many Python built-in functions
+- C functions
+
+Below is an example of a graph break due to the function ``copy.deepcopy`` from a Python builtin library
+(exact output may differ).
+
+.. code-block:: py
+
+    import copy
+    import torch
+
+    @torch.compile
+    def fn(x):
+        x = x + 1
+        y = copy.deepcopy(x)
```

Comment:
This is not a great example, because in principle we can eventually support this builtin

Reply:
Switched to using `eval`

Reply:
eval is a bad example too, eval is actually easy to support when we do bytecode interpretation LOL

Reply:
What about `open`?

Reply:
that one's good!